### ISSUE TITLE
[interoperability] add flag to source when we know it's provided by geoserver

### ISSUE THREAD

**Issue by landryb** [geosolutions-it/MapStore2#8167]
## Description

Right now, many features in mapstore2 sadly assume that the remote is geoserver (because the feature was developed and tested only against geoserver) - and it also assumes that the remote geoserver is located at `/geoserver/` which might not be the case (cf https://github.com/geosolutions-it/MapStore2/blob/master/web/client/utils/LayersUtils.js#L25)

Consequently, there are many small interoperability issues when using layers coming from non-geoserver sources (eg mapserver, mapproxy, qgis server..).
- #7809
- #7877
- #7879
- georchestra/mapstore2-georchestra#300
- georchestra/mapstore2-georchestra#301
- georchestra/mapstore2-georchestra#302

all those issues could be 'easily' solved if in the catalog/layer object there was a boolean flag which said if the remote was geoserver - the corresponding printing/loading code could then be amended to behave differently depending on the remote.

**What kind of improvement you want to add?** (check one with "x", remove the others)
- [x] Minor changes to existing features

## Other useful information

implementation ideas (from https://github.com/geosolutions-it/MapStore2/pull/7811#issuecomment-1047998536):
- parse `GetCapabilities` and look for `GEOSERVER`in `/Service/KeywordList/Keyword` but that assumes the admin doesnt change server keywords
- look for `updateSequence` attribute on the toplevel `WMS_Capabilities` XML element of `GetCapabilities`
- have a checkbox in the UI (checked by default, with _remote is geoserver_ label) in the advanced catalog settings, which would set the flag on the catalog js object

**Comment by allyoucanmap** [comment]
> and it also assumes that the remote geoserver is located at /geoserver/ which might not be the case (cf https://github.com/geosolutions-it/MapStore2/blob/master/web/client/utils/LayersUtils.js#L25)

Took a look to the use of the geoserver regex

- [getRegGeoserverRule](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L664)  used only inside [findGeoServerName](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L672) inside the same file
- [findGeoServerName](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L672) is used by following components:
  - [styleeditor](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/epics/styleeditor.js#L257) used to initialize the rest api request for style editor. If it does not match it will fallback to a normal get capabilities without activate the style editor functionality
  - [getCapabilitiesUrl](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L684) used to make an optimization on the capabilities request to target the layer workspace and reduce the size of the response. If the layer does not match the regex it will fallback to a normal get capabilities

Additionally the utils allows to set the regex with [setRegGeoserverRule](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L658) inside a downstream project to extend the possible supported GeoServer. [Here an example](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L658) of use case of this property in geonode.

We saw that MapStore provides fallbacks for all the cases based on the previous investigation and it expose a utils that allows to extend the supported path in the regex.
My suggestion is to move this possible improvement to a different issue if we intend to find a different approach the geoserver url detection cc @tdipisa

**Comment by tdipisa** [comment]
@allyoucanmap 

> geoserver the wms request will works as now (geoserver)

I agree to include also this now

> - [WANTDO][UI] Include in the WMS layer [general setting panel](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/components/TOC/fragments/settings/General.jsx) a select input to allow a user to select the server type of the layer. This implementation should also include an localConfig option to hide it

> My suggestion is to move this possible improvement to a different issue if we intend to find a different approach the geoserver url detection cc @tdipisa

I totally agree @allyoucanmap. Can you please open the new issue including the possibility to make it configurable from Catalog advanced settings? You can assign the issue to me for the moment.

**Comment by landryb** [comment]
> Additionally the utils allows to set the regex with [setRegGeoserverRule](https://github.com/geosolutions-it/MapStore2/blob/c6fad76296278e5833e8eca8443fbdb029bf4c99/web/client/utils/LayersUtils.js#L658) inside a downstream project to extend the possible supported GeoServer. [Here an example](https://github.com/GeoNode/geonode-mapstore-client/blob/master/geonode_mapstore_client/client/js/utils/AppUtils.js#L72) of use case of this property in geonode.

Oh great, nice finding ! That'll allow me to override it from ms2-geor, i'll play with that. Maybe calling `setRegGeoserverRule` here: https://github.com/georchestra/mapstore2-georchestra/blob/master/js/app.jsx#L165 ?

**Comment by allyoucanmap** [comment]
> Oh great, nice finding ! That'll allow me to override it from ms2-geor, i'll play with that. Maybe calling setRegGeoserverRule here: https://github.com/georchestra/mapstore2-georchestra/blob/master/js/app.jsx#L165 ?

Yes, I think that's the correct place. It should be initialized before the main application start

**Comment by landryb** [comment]
i've started a first crude wip in https://github.com/landryb/MapStore2/commit/0a5e489cac9250df23a467976314a9e02064c9b4 directly on ms2 (ie outside of ms2-georchestra project) but it blows when loading http://localhost:8080/mapstore/?debug=true.
complaining that `options isnt defined`. I guess so , but i have some interrogations:
- how am i supposed to have the layer object (and access the `serverType` prop) in the 3 callers of `getWMSVendorParams` ? To my understanding i only have access to `options` at that point.. for example should i move `wmsToOpenlayersOptions` call to after the layer object creation in https://github.com/geosolutions-it/MapStore2/blob/master/web/client/components/map/openlayers/plugins/WMSLayer.js#L193 and pass layer object to it so that it gets down to `getWMSVendorParams` ?
- doing a regular build via `npm run fe:build` or `npm start`, even with `?debug=true` i dont have the full source map in the browser debugging console - eg only access to the minified source. Something already recently reported in georchestra/mapstore2-georchestra#568 ? something else than https://docs.mapstore.geosolutionsgroup.com/en/latest/developer-guide/developing/#debugging-the-frontend to do ? I havent touched the webpack config and i'm not supposed to do so, as prod defaults to false
- There's already `optionsToVendorParams` in https://github.com/geosolutions-it/MapStore2/blob/master/web/client/utils/VendorParamsUtils.js but it only seems to be about CQL filters, should my new function move there instead ?

Will keep digging.

**Comment by landryb** [comment]
still digging, i think i have a better understanding of the bits here and there at stake - the "options" object should have all the bits i need, i think. Will be able to resume work on this from next wednesday.

**Comment by landryb** [comment]
@allyoucanmap what i have in https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167 is hacky and doesnt work so far. Some issues:
- in https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-18c0efc001f84f53e277264fbd7753558f932eaca95a019d7803b0140cc664f3R78 `record` object doesnt have my `serverType` property.
- hiding the 'use cache options' section with https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-b07795533c0bae745f584d5c09b87684ef445551a63c859d243b030e4cfccd38R222 doesnt seem to work
- i cant set the 'default value' for serverType, ie setting a value for a server in `localConfig.json` doesn't set the value for the selectbox in the UI. And selecting `GEOSERVER` or `NO_VENDOR` in the selectbox doesnt seem persisted in the UI (but i do see the  `CHANGE_SERVICE_PROPERTY` action:
```
Object { type: "CATALOG:CHANGE_SERVICE_PROPERTY", property: "serverType", value: "GEOSERVER" }
```
 - i guess my changes in https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-8bb020509f0b81ce10ff349328dab6db83acad1349c3f1b58d2f48edbad3be64R185 are probably wrongly wired to the object
- my understanding is that the constants (eg `GEOSERVER`, `NO_VENDOR`) are used throught the code, and the 'values' (eg `geoserver` & `no-vendor`) are the ones saved in the objects/services ?
- the `layer` object (coming from 'options') in https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-8a8dd5af00d8f940d51057bcc4da099ebfd1be9f10a0ec4a75e0270834d36ebcR844 has `undefined` for `serverType` (so the value coming from the `catalog` object doesn't reach it ?). Also by default it doesn't have `singleTile` or `tiled` properties (even undefined), but it has them once i change them from the defaults via the UI in the layer properties. I'm not 100% certain i'm accessing the right object ?

more digging to come today.

**Comment by allyoucanmap** [comment]
> in https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-18c0efc001f84f53e277264fbd7753558f932eaca95a019d7803b0140cc664f3R78 record object doesnt have my serverType property.

@landryb I would suggest to take a look to this PR https://github.com/geosolutions-it/MapStore2/pull/9022 for the part concerning the RasterAdvancedSettings.js component where to correctly apply the property to the layer I used:
```js
onChangeServiceProperty("layerOptions", { ...service.layerOptions, singleTile: e.target.checked })
```

in this case should be

```js
onChangeServiceProperty("layerOptions", { ...service.layerOptions, serverType: event?.value })
```
If you add the property in the layerOptions property it should be directly included in the added layer

**Comment by landryb** [comment]
ok, found out why serverType didnt percolate to the layer object - 3d3db6ce2 helps and my layer object has the proper `serverType` value from the catalog service.

**Comment by allyoucanmap** [comment]
> thanks, i've looked at https://github.com/geosolutions-it/MapStore2/pull/9022 and it seems much simpler than what i'm doing, so should i put the serverType property directly inside the existing layerOptions object instead of adding it as a new layer property passing it via the server record/catalog ? or that's still needed to set its initial value ?

Yes, correct, you should put directly serverType inside the layerOptions object without adding a new property. While I was working in the PR I noticed that the layerOptions property was used for the [tileSize layer property](https://github.com/geosolutions-it/MapStore2/blob/6d2307abc85416d4dcccac3e4c758acac2a5b3e6/web/client/components/catalog/editor/AdvancedSettings/RasterAdvancedSettings.js#L162-L172) and then the content of the [layerOptions](https://github.com/geosolutions-it/MapStore2/blob/6d2307abc85416d4dcccac3e4c758acac2a5b3e6/web/client/api/catalog/WMS.js#L153-L156) added directly [to the layer](https://github.com/geosolutions-it/MapStore2/blob/6d2307abc85416d4dcccac3e4c758acac2a5b3e6/web/client/api/catalog/WMS.js#L106). To keep it consistent I added also the singleTile layer property and I think we should do the same for serverType.

> i guess i should also rebase on top of your branch so that all onChangeServiceProperty calls for layeroptions have the ...service.layerOptions bit ?

The PR branch is not on master yet so I would simply suggest to add this line change directly on your code without rebase https://github.com/geosolutions-it/MapStore2/pull/9022/files#diff-8bb020509f0b81ce10ff349328dab6db83acad1349c3f1b58d2f48edbad3be64R179

**Comment by landryb** [comment]
@allyoucanmap i dont think this issue was closed by #7865, rather georchestra/mapstore2-georchestra#300, and only when it's upgraded to 2023.02.

**Comment by allyoucanmap** [comment]
@landryb it was automatically closed because the P https://github.com/geosolutions-it/MapStore2/pull/7865 was merged, just reopened thanks

**Comment by allyoucanmap** [comment]
@landryb here the options are created only with the keys https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167#diff-8bb020509f0b81ce10ff349328dab6db83acad1349c3f1b58d2f48edbad3be64R80. If we want to use the value we should create the options in a different way, this an example:

```js
const getServerTypeOptions = () => {
    return Object.keys(ServerTypes).map((key) => ({ label: key, value: ServerTypes[key] }));
};
```

**Comment by landryb** [comment]
i've tried applying the logic to the layers added in map of a new dashboard, using:
```
--- a/web/client/plugins/widgetbuilder/enhancers/layerSelector.js
+++ b/web/client/plugins/widgetbuilder/enhancers/layerSelector.js
@@ -16,7 +16,7 @@ import API from '../../../api/catalog';
 export const toLayer = (r, service) => ["tms", "wfs"].includes(service?.type) // for tms and wfs the layer is ready
     ? r
     // the type wms is default (for csw and wms), wmts have to be passed. // TODO: improve and centralize more
-    : API[service?.type || 'wms'].getLayerFromRecord(r);
+    : API[service?.type || 'wms'].getLayerFromRecord(r, service);

 // checks for tms wmts in order to addSearch() to skip addSearch
 export const addSearchObservable = (selected, service) => ["tms", "wmts"].includes(service?.type) ? Rx.Observable.of(toLayer(selected, service)) : addSearch(toLayer(selected, service));
```

but that blows because i try to use `service.serverType` while `service` is undef when adding a layer to a dashboard:
```
TypeError: service is undefined
    recordToLayer webpack://mapstore2/./web/client/api/catalog/WMS.js?:140
    getLayerFromRecord webpack://mapstore2/./web/client/api/catalog/WMS.js?:237
    toLayer webpack://mapstore2/./web/client/plugins/widgetbuilder/enhancers/layerSelector.js?:32
```

i've thought a bit more about moving `serverType` to `layerOptions` completely (like `singleTile` ?) but i'm not 100% sure it'll be properly saved in the map when exporting it.. afaict there's no `layerOptions` in `saveLayer` while there's `singleTile`.. a bit confused now.

**Comment by landryb** [comment]
i *think* i have something that works both for maps & dashboards. @allyoucanmap  would love feedback on https://github.com/geosolutions-it/MapStore2/compare/master...landryb:MapStore2:wip/8167 :)

so far i've tested:
- adding layers to a map without vendor params from a server configured from localConfig.json (eg `"layerOptions": {"serverType": "no-vendor" },`)
- adding layers to a map without vendor params from a server configured as NO_VENDOR via the UI
- exporting and importing the map, the vendor params are not sent so serverType is properly persisted in `map.json`
- adding layers to a dashboard without vendor params from a server configured as NO_VENDOR via the UI
- exporting and importing the dashboard, the vendor params are not sent so serverType is properly persisted in `dashboard.json`

selecting `GEOSERVER` in the selector still doesn't trigger an action nor a change in the UI.

**Comment by landryb** [comment]
> @landryb I've updated the description including a tasklist and defining missing issues MS side. I would like to assign them to you so that I've added you in the collaborators list with read permissions for the purpose. Can you please accept the invitation? Then, you will be able to put your comments in the related issue so that it will be also better for the testing phase. We can continue using this issue as an Epic one just to collect together involved issues.

I think i was already able to add comments to those issues as a regular contributor, but thanks for adding me to the collaborators team.

Regarding the task list, i'm a little confused by the concept because sure, all those issues are 'interoperability-related', but only #7809, #7877 and #7879 will be fixed by this issue.

For the others:
- #9030 should be fixed now that #7865 is merged in master - i've tested it throroughly (cf https://github.com/geosolutions-it/MapStore2/pull/7865#issuecomment-1463580373) and from what i could understand @ElenaGallo is working on validating/testing the fixes.
- as discussed in our last meeting, #9031 & #9007 wont get fixed by this issue, work from your team is needed.. unless i got it wrong ?

**Comment by tdipisa** [comment]
@landryb 

>I think i was already able to add comments to those issues as a regular contributor, but thanks for adding me to the collaborators team.

I was not able to assign issues not created by you without including you in the collaborators list.

>Regarding the task list, i'm a little confused by the concept because sure, all those issues are 'interoperability-related', but only https://github.com/geosolutions-it/MapStore2/issues/7809, https://github.com/geosolutions-it/MapStore2/issues/7877 and https://github.com/geosolutions-it/MapStore2/issues/7879 will be fixed by this issue.

The Task List I've added in description is only a way we are using to track all issue involved in a macro task that is including more than one issue (what we call an _Epic_ indeed). When you have created this issue you linked in description all existing interoperability issues together. Generally speaking, the idea here is to have a dedicated issue for each point in Task List and a final PR connected for each. This is the procedure we usually follow for better manage code reviews and functional tests also in our board.

**Comment by landryb** [comment]
i've rebased my branch and added some unit tests (fa787e65, 8870c0e3, a06931ef) for the new functions but i havent figured out how to test that a catalog entry with `serverType` set to something yielded layer objects with `layerOptions.serverType` set to the corresponding value. My guess is that it should go into https://github.com/geosolutions-it/MapStore2/blob/master/web/client/api/catalog/__tests__/WMS-test.js#L10.

Trying to run the existing tests in that file, all the _Test correctness of the WMS APIs_ tests fail so i have probably something wrong somewhere:
```
  Test correctness of the WMS APIs
    � wms limited srs
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:69:75)

    � wms multiple urls
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:81:75)

    � wms layer options
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:97:75)

    � wms layer with visibility limits
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:112:75)

    � wms with no ogcServiceReference.url
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:131:75)

    � wms layer with formats
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:212:75)

    � wms layer with force proxy
      Chrome Headless 111.0.5563.64 (Linux x86_64)
    TypeError: Cannot read properties of undefined (reading 'serverType')
        at recordToLayer (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2489:25)
        at getLayerFromRecord (webpack://mapstore2/./web/client/api/catalog/WMS.js?:2616:44)
        at Context.eval (webpack://mapstore2/./web/client/api/catalog/__tests__/WMS-test.js?:228:75)
```

i have this wip to test the `serverType` layerOption like `tileSize`:
```
--- a/web/client/api/catalog/__tests__/WMS-test.js
+++ b/web/client/api/catalog/__tests__/WMS-test.js
@@ -90,12 +90,14 @@ describe('Test correctness of the WMS APIs', () => {
         }, {
             url: 'http://sample',
             layerOptions: {
-                tileSize: 512
+                tileSize: 512,
+                serverType: "no-vendor"
             }
         });
         expect(records.length).toBe(1);
         const layer = getLayerFromRecord(records[0]);
         expect(layer.tileSize).toBe(512);
+        expect(layer.serverType).toBe("no-vendor");
     });
```

maybe there should be a test for `singleTile` added in #9022 ?

**Comment by landryb** [comment]
failing tests fixed with 29efec41, and new test added in faf6575. Can probably be improved :)

**Comment by allyoucanmap** [comment]
> @allyoucanmap : with [9057bca](https://github.com/geosolutions-it/MapStore2/commit/9057bca2ada0125e3194b9d14f03916939f43327) which removes the "hardcoded" default to _'GEOSERVER'_, in my testing with the default csw (which has no `serverType` set in config), the advanced options shows 'Select...' (so the value is undefined) and from that point i can only select _'NO_VENDOR'_, i can't select _'GEOSERVER'_. Strange ?

> That strange behaviour doesnt reproduce with the two WMS entries where i've pre-set serverType to either "geoserver" or "no-vendor", for both the default value is correct and i can select the other value.

@landryb this seems a conflict with the first row of the codemirror library underneat for the Static Filter field available only in CSW, also other select input have similar problem when the options position match the filter line

![image](https://user-images.githubusercontent.com/19175505/227163044-af81c124-bbea-4139-aaad-560c168a085e.png)

**Comment by allyoucanmap** [comment]
@landryb to solve this https://github.com/geosolutions-it/MapStore2/issues/8167#issuecomment-1480878868 we need to apply a style to the [CSW filter container](https://github.com/landryb/MapStore2/blob/feat/8167/web/client/components/catalog/editor/AdvancedSettings/CSWFilters.jsx#L119) with the following values `{ position: 'relative', zIndex: 0 }` to ensure the codemirror editor does not go over the select options

**Comment by landryb** [comment]
> @landryb to solve this [#8167 (comment)](https://github.com/geosolutions-it/MapStore2/issues/8167#issuecomment-1480878868) we need to apply a style to the [CSW filter container](https://github.com/landryb/MapStore2/blob/feat/8167/web/client/components/catalog/editor/AdvancedSettings/CSWFilters.jsx#L119) with the following values `{ position: 'relative', zIndex: 0 }` to ensure the codemirror editor does not go over the select options

right - hadn't thought about it, that was indeed an issue only for CSW catalogs and not for WMS catalogs, tested with 986ea7592a4ebbd31593f239d8ae92105818d478 and that bit is now working fine. Thanks !
